Pine - Lety P. and Karina G.

Infernormal opened the pull request.

CheezItMan commented on app/models/customer.py:
```diff
@@ -1,4 +1,21 @@
+    registered_at = db.Column(db.DateTime,nullable=True)
+    postal_code = db.Column(db.String)
+    phone = db.Column(db.String)
+    videos = db.relationship("Video", secondary = "Rentals", backref = "Customers",passive_deletes=True)
```
👍

CheezItMan commented on app/models/rental.py:
```diff
@@ -1,4 +1,8 @@
 from app import db
 
 class Rental(db.Model):
```
No helper methods here?

CheezItMan commented on app/models/video.py:
```diff
@@ -1,4 +1,21 @@
+    title = db.Column(db.String)
+    release_date = db.Column(db.DateTime)
+    #available_inventory should be an attribute of the video
+    available_inventory = db.Column(db.Integer)
```
I just want to point out that the available inventory can be calculated by taking the total inventory and subtracting the number of rentals that are still outstanding.

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+def create_videos():
+    if request.method == "POST":
+        request_body = request.get_json()
+        input_check = validate_video_input(request_body)
```
Great use of a helper function.  

Just a note, that doing this check and the following if-statement is a bit awkward.  Another way to do this is with a decorator which can dry this up a bit. Not a problem for this project, but something to note for future learning.

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+def handle_customers():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "name" not in request_body:
+            return ({
+                "details": "Request body must include name."
+            }),400
+        elif "phone" not in request_body:
+            return ({
+                "details": "Request body must include phone."
+            }),400
+        elif "postal_code" not in request_body:
+            return ({
+                "details": "Request body must include postal_code."
+            }),400
```
I just want to note how this validation section makes the function VERY long.  Breaking it out with a helper function or a decorator would be ideal.

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+
+        db.session.commit()
+
+        return ({
+            "name": f"{customer.name}",
+            "phone": f"{customer.phone}",
+            "postal_code": f"{customer.postal_code}"
```
It would be better to do this with a helper method in the `Customer` model

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+def handle_checkout():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "customer_id" not in request_body:
+            return ({
+                "details": "Request body must include customer_id."
+            }),400
+        elif "video_id" not in request_body:
+            return ({
+                "details": "Request body must include video_id."
+            }),400
```
Good place for a validator function

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+        return ({
+            "customer_id": new_rental.customer_id,
+            "video_id": new_rental.video_id,
+            "due_date": datetime.datetime.today() - datetime.timedelta(days=7),
+            "videos_checked_out_count": videos_checked_out_count(new_rental),
+            "available_inventory": available_inventory - rentals_associated(new_rental)
```
1. I like the calculations done here in the dict.
2. This would be great as a `Rental` helper method (maybe `to_dict`).

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+
+#HELPER FUNCTIONS
+
+def rentals_associated(rental):
+    return len(db.session.query(Rental).filter(Rental.video_id == rental.video_id).all())
```
This might be better done with a relationship, so you can do `video.rentals`. Or `rental.video.rentals` to get all the other rentals.

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+    return len(db.session.query(Rental).filter(Rental.video_id == rental.video_id).all())
+
+
+def videos_checked_out_count(rental):
+    return len(db.session.query(Rental).filter(Rental.customer_id == rental.customer_id).all())
```
Good helper function, but this would be better as a helper method in the Video class.

CheezItMan commented on app/routes.py:
```diff
@@ -0,0 +1,291 @@
+    return len(db.session.query(Rental).filter(Rental.customer_id == rental.customer_id).all())
+
+
+def validate_video_input(request_body):
```
Good idea with the helper method.

CheezItMan commented on app/models/video.py:
```diff
@@ -1,4 +1,21 @@
 from app import db
```
I just want to note that this file is almost 400 lines long.  Maybe consider breaking it up into separate files.

CheezItMan reviewed (commented):
Nicely done, you hit the learning goals Lety & Karina.  I left some comments, mostly on organization and better use of relationships.  This is a pretty good submission and does everything required.  🟢

Infernormal commented:
@CheezItMan even though we got green for this project I felt like I need to refactor it, I addressed most of the feedback, hopefully it looks better now :)
